bigtable-hbase-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/adapters/CheckAndMutateUtil.java

```diff
@@ -71,67 +70,129 @@ public static boolean wasMutationApplied(
+
+    /**
+     * <p>
+     * RequestBuilder.
+     * </p>
+     * @param hbaseAdapter a {@link HBaseRequestAdapter} used to convert HBase
+     *          {@link org.apache.hadoop.hbase.client.Mutation} to Cloud Bigtable {@link Mutation}
+     * @param row the RowKey in which to to check value matching
+     * @param family the family in which to check value matching.
+     */
+    public RequestBuilder(HBaseRequestAdapter hbaseAdapter, byte[] row, byte[] family) {
+      requestBuilder.setRowKey(ByteString.copyFrom(row));
+      this.row = Preconditions.checkNotNull(row, "row is null");
+      this.family = Preconditions.checkNotNull(family, "family is null");
+
+      // TODO (issue #1709): The hbaseAdapter used here should not set client-side timestamps.
+      this.hbaseAdapter = hbaseAdapter;
+      requestBuilder.setTableName(hbaseAdapter.getBigtableTableName().toString());
     }
 
-    CheckAndMutateRowRequest.Builder requestBuilder =
-        CheckAndMutateRowRequest.newBuilder();
+    public RequestBuilder qualifier(byte[] qualifier) {
+      this.qualifier = Preconditions.checkNotNull(qualifier, "qualifier is null. Consider using" +
+          " an empty byte array, or just do not call this method if you want a null qualifier");
+      return this;
+    }
 
-    requestBuilder.setTableName(hbaseAdapter.getBigtableTableName().toString());
-
-    requestBuilder.setRowKey(ByteString.copyFrom(row));
-    Scan scan = new Scan().addColumn(family, qualifier);
-    scan.setMaxVersions(1);
-    if (value == null) {
-      // If we don't have a value and we are doing CompareOp.EQUAL, we want to mutate if there
-      // is no cell with the qualifier. If we are doing CompareOp.NOT_EQUAL, we want to mutate
-      // if there is any cell. We don't actually want an extra filter for either of these cases,
-      // but we do need to invert the compare op.
-      if (CompareFilter.CompareOp.EQUAL.equals(compareOp)) {
-        requestBuilder.addAllFalseMutations(mutations);
-      } else if (CompareFilter.CompareOp.NOT_EQUAL.equals(compareOp)) {
+    public RequestBuilder ifNotExists() {
+      Preconditions.checkState(compareOp == null,
+          "ifNotExists and ifMatches are mutually exclusive");
+      this.checkNonExistence = true;
+      return this;
+    }
+
+    public RequestBuilder ifEquals(CompareOp compareOp, byte[] value) {
+      return ifMatches(CompareOp.EQUAL, value);
+    }
+
+    public RequestBuilder ifMatches(CompareOp compareOp, byte[] value) {
+      Preconditions.checkState(checkNonExistence == false,
+          "ifNotExists and ifMatches are mutually exclusive");
+
+      this.compareOp = Preconditions.checkNotNull(compareOp, "compareOp is null");
+
+      // TODO (issue #1704): only NOT_EQUALS should allow null.  Everything else should use ifNotExists();
+      this.value = value;
+      return this;
+    }
+
+    public CheckAndMutateRowRequest buildForPut(Put put) throws DoNotRetryIOException {
```

Comment:
It might be better to split the 2 responsibilities:
`addPut` and `build()`

Reply:
ack